x-pack/plugins/fleet/public/applications/fleet/sections/agent_policy/create_package_policy_page/components/steps/components/dataset_component.tsx

```diff
@@ -0,0 +1,187 @@
+
+const FormRow = styled(EuiFormRow)`
+  .euiFormRow__label {
+    flex: 1;
+  }
+
+  .euiFormRow__fieldWrapper > .euiPanel {
+    padding: ${(props) => props.theme.eui?.euiSizeXS};
+  }
+`;
+
+interface SelectedDataset {
+  dataset: string;
+  package: string;
+}
+
+export const DatasetComponent: React.FC<{
+  value?: SelectedDataset | string;
+  onChange: (newValue: SelectedDataset) => void;
+  datastreams: DataStream[];
+  pkgName?: string;
+  isDisabled?: boolean;
+  fieldLabel: string;
+  description?: string;
+}> = ({ value, onChange, datastreams, isDisabled, pkgName = '', fieldLabel, description }) => {
+  const datasetOptions =
+    datastreams.map((datastream: DataStream) => ({
+      label: datastream.dataset,
+      value: datastream,
+    })) ?? [];
+  const existingGenericStream = datasetOptions.find((ds) => ds.label === GENERIC_DATASET_NAME);
+  const valueAsOption = value
+    ? typeof value === 'string'
+      ? { label: value, value: { dataset: value, package: pkgName } }
+      : { label: value.dataset, value: { dataset: value.dataset, package: value.package } }
+    : undefined;
+  const defaultOption = valueAsOption ||
+    existingGenericStream || {
+      label: GENERIC_DATASET_NAME,
+      value: { dataset: GENERIC_DATASET_NAME, package: pkgName },
+    };
+
+  const [selectedOptions, setSelectedOptions] = useState<Array<{ label: string }>>([defaultOption]);
+  const [isInvalid, setIsInvalid] = useState<boolean>(false);
+  const [error, setError] = useState<string | undefined>(undefined);
+
+  useEffect(() => {
+    if (!value || typeof value === 'string') onChange(defaultOption.value as SelectedDataset);
+  }, [value, defaultOption.value, onChange, pkgName]);
+
+  const onDatasetChange = (newSelectedOptions: Array<{ label: string; value?: DataStream }>) => {
+    setSelectedOptions(newSelectedOptions);
+    const dataStream = newSelectedOptions[0].value;
+    const invalid = !newSelectedOptions[0].label.match(DATASET_REGEX);
+    if (invalid) {
+      setIsInvalid(true);
+      setError(REGEX_ERROR);
+    } else {
+      setIsInvalid(false);
+      setError(undefined);
+    }
+    onChange({
+      dataset: newSelectedOptions[0].label,
+      package: !dataStream || typeof dataStream === 'string' ? pkgName : dataStream.package,
+    });
+  };
+
+  const onCreateOption = (searchValue: string = '') => {
+    const normalizedSearchValue = searchValue.trim().toLowerCase();
+    if (!normalizedSearchValue) {
+      return;
+    }
+    const newOption = {
+      label: searchValue,
+      value: { dataset: searchValue, package: pkgName },
+    };
+    setSelectedOptions([newOption]);
+    const invalid = !searchValue.match(DATASET_REGEX);
+    if (invalid) {
+      setIsInvalid(true);
+      setError(REGEX_ERROR);
+    } else {
+      setIsInvalid(false);
+      setError(undefined);
+    }
+    onChange({
+      dataset: searchValue,
+      package: pkgName,
+    });
+  };
+  return (
+    <FormRow
```

Comment:
I did a rename of `DatasetComboBox` to `DatasetComponent` because it made sense to have the combo box component together with `FormRow` in one component. I think it's more readable this way than passing the `isInvalid` and `error` field up in the components chain to show the dataset validation error.